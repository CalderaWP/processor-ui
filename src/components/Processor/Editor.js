import React from 'react';
import PropTypes from 'prop-types';
import * as CalderaComponents from '@caldera-labs/components';
import classNames from 'classnames'
export class Editor extends React.PureComponent {
	constructor(props) {
		super(props);
		this.state = {
			opacity: 1,
			hovered: false,
		};
		this.mouseEnter = this.mouseEnter.bind(this);
		this.mouseLeave = this.mouseLeave.bind(this);
		this.changeType = this.changeType.bind(this);
		this.handleChange = this.handleChange.bind(this);
		this.configFields = this.configFields.bind(this);
		this.labelAs = this.labelAs.bind(this);

	}

	/**
	 * Handler for the mouse enter event
	 */
	mouseEnter() {
		this.setState({hovered: true})
	}

	/**
	 * Handler for the mouse leave event
	 */
	mouseLeave() {
		this.setState({hovered: false})
	}

	/**
	 * Prepares config fields for CalderaComponents.RenderGroup
	 *
	 * @return {Array}
	 */
	configFields(){
		if( Array.isArray( this.props.configFields ) ) {
			return this.props.configFields;
		}

		let fields = [];
		if ( 'object' === typeof this.props.configFields ) {
			Object.keys(this.props.configFields).forEach(configFieldId => {
				fields[configFieldId] = this.props.configFields[configFieldId];
				fields[configFieldId].onValueChange = (newValue) => {
					this.handleChange(configFieldId, newValue);
				};
			});
		}
		return fields;
	}

	/**
	 * Updates the processor config fields and dispatches the change handler
	 *
	 * @param {String} configFieldId Id of configuration field to update.
	 * @param {String|Array|Number} newValue The new value
	 */
	handleChange(configFieldId,newValue){
		let configFields = this.props.configFields;
		configFields[configFieldId] = {
			...configFields[configFieldId],
			value:newValue
		};
		this.props.onUpdateProcessor({
			ID: this.props.ID,
			type: this.props.type,
			label: this.props.label,
			configFields
		});
	}

	/**
	 * Dispatch a change in processor type
	 *
	 * @param {*} event
	 */
	changeType(event){
		this.props.onUpdateProcessor({
			ID: this.props.ID,
			label: this.props.label,
			type: event.target.value,
			configFields: this.props.configFields
		});
	}

	/**
	 * Find the label to show for a processor
	 * @return {string}
	 */
	labelAs(){
		return this.props.label
			? this.props.label
			: this.props.type
				? `${this.props.type} - ${this.props.ID}`
				: this.props.ID;
	}


	render() {
		let removeStyle = {
			//opacity: 0.1
			opacity: 1
		};
		if ( this.state.hovered){
			removeStyle.opacity = 1
		}

		const options = [
			{
				label: 'Email',
				value: 'email'
			},
			{
				label: 'Redirect',
				value: 'redirect'
			}

		];
		return (
			<div
				className={ classNames(this.props.className, {
						'processor-has-type' : 'string' === typeof this.props.type,
						'caldera-forms-processor' : true,
						'not-opened': !this.props.isOpen,
						'is-opened': this.props.isOpen
					})
				}
				//onMouseEnter={this.mouseEnter}
				//onMouseLeave={this.mouseLeave}
			>
				<div className={'processor-label'}>{this.labelAs()}</div>
				<div>
					Type: <span className={'processor-type'}>{this.props.type}</span>
				</div>
				<select
					className={'processor-type-chooser'}
					onChange={this.changeType}
					value={this.props.type}
				>
					<option/>
					{options.map(option => {
							return (
								<option
									key={option.value}
									value={option.value}
								>
									{option.label}
								</option>
							);
						}
					)}
				</select>

				{this.props.type &&
					<div
						className={'processor-editor'}
					>
						<CalderaComponents.RenderGroup
							className={'caldera-forms-processor-config'}
							configFields={this.configFields()}
						/>
					</div>
				}

			</div>
		)
	}
}

/**
 * Prop definitions for Editor Property
 * @type {{processor: *, onUpdateProcessor: *, form: *, configFields: shim}}
 */
Editor.propTypes = {
	onUpdateProcessor: PropTypes.func.isRequired,
	form: PropTypes.object.isRequired,
	configFields: PropTypes.oneOfType([
			PropTypes.array,
			PropTypes.object
	]

	),
	isOpen: PropTypes.bool,
	ID: PropTypes.string.isRequired,
	type: PropTypes.string,
	label: PropTypes.string,

};

Editor.defaultProps = {
	isOpen: false,
	configFields: {}
};