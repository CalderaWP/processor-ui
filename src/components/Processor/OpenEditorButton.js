import React from 'react';
import PropTypes from 'prop-types';

/**
 * Displays button to open or close processor editor
 *
 * @param {Object} props
 * @returns {*}
 * @constructor
 */
export const OpenEditorButton = (props) => {
	return (

			<button
				style={props.style}
				className={props.className}
				onClick={props.onClick}
			>
				{!props.isOpen &&
					<span>{props.labelToOpen}</span>
				}
				{props.isOpen &&
					<span>{props.labelToClose}</span>
				}
			</button>

	)

};

/**
 * Props requirements for OpenEditorButton component
 *
 * @type {{isOpen: *, labelToOpen: shim, labelToClose: shim, onClick: *}}
 */
OpenEditorButton.propTypes = {
	isOpen: PropTypes.bool.isRequired,
	labelToOpen: PropTypes.string,
	labelToClose: PropTypes.string,
	onClick: PropTypes.func.isRequired,
};

/**
 * Default props for OpenEditorButton
 *
 * @type {{labelToOpen: string, labelToClose: string}}
 */
OpenEditorButton.defaultProps = {
	labelToOpen: 'Edit Settings',
	labelToClose: 'Close Editor',
};
