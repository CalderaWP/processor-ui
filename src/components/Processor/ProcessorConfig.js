import React from 'react';
import PropTypes from 'prop-types';

/**
 * Displays button to open or close processor editor
 *
 * @param {Object} props
 * @returns {*}
 * @constructor
 */
export const ProcessorConfig = (props) => {

	return(
		<div
			className={props.className}
		>
			{Array.from( props.configFields.values() ).map( (field,i)=> {


			})}
		</div>
	)


};


/**
 * Props requirements for ProcessorConfig component
 *
 */
ProcessorConfig.propTypes = {
	configFields: PropTypes.object.isRequired,
	onUpdate: PropTypes.func.isRequired,
};

/**
 * Default props for ProcessorConfig
 *
 */
ProcessorConfig.defaultProps = {
	configFieldsDefaults: {},
	configValues: new Map(),
};
