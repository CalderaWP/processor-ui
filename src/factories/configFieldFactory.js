import {
	getConfigFieldValueOrDefault,
	getConfigFieldDefaultValue} from "./util";

/**
 * Factory for single processor configs
 *
 * @param {Object} configField Config fields for processor config field.
 * @param {Object} configFieldDefaults Optional. Config field defaults for processor config field.
 * @returns {{}}
 */
export const configFieldFactory = (configField, configFieldDefaults = {} ) => {
	configField.value = getConfigFieldValueOrDefault(configField);
	configField.default = getConfigFieldDefaultValue(configField);
	configFieldDefaults.default = getConfigFieldDefaultValue(configFieldDefaults);
	if( null === configField.value ){
		configField.value = configFieldDefaults.default;
	}
	return {
		...configFieldDefaults,
		...configField,
	}
};
