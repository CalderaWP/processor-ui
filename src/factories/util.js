/**
 * Finds value property of config field or config field default
 * @param {Object} configField Config field to search for default in
 * @returns {String|Number|Array|null}
 */
export const getConfigFieldValue = (configField) => {
	return 'object'=== typeof configField && configField.hasOwnProperty('value')
		? configField.value
		: null;
};

/**
 * Finds default property of config field or config field default
 * @param {Object} configField Config field to search for default in
 * @returns {String|Number|Array|null}
 */
export const getConfigFieldDefaultValue = (configField) => {
	return 'object'=== typeof configField && configField.hasOwnProperty('default')
		? configField.default
		: null;
};

/**
 * Finds value of config field, falling back to default value
 * @param {Object} configField Config field to search for value or defualt in
 * @returns {String|Number|Array|null}
 */
export const getConfigFieldValueOrDefault = (configField ) => {
	return null !== getConfigFieldValue(configField)
		? getConfigFieldValue(configField)
		: null !== getConfigFieldDefaultValue(configField)
			? getConfigFieldDefaultValue(configField)
			: null;
};

/**
 * Convert an object (should have meaningful indexes) to a map
 * @param {Object} object Object to convert
 * @return {Map<any, any>}
 */
export const objectToMap = (object) => {
	const map = new Map();
	Object.keys(object).forEach(key => {
		map.set(key, object[key]);
	});
	return map;
}