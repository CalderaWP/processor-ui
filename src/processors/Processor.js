/**
 * Base class for processors
 *
 * Should be an interface or type
 */
export class Processor {

	/**
	 * Create new processor
	 * @param {Object} configFields
	 * @param {Object} configValues
	 */
	constructor(configFields = {}, configValues = {} ){
		this.configFields = configFields;
		this.configValues = configValues;
	}

	/**
	 * Get the config fields
	 *
	 * @returns {Object}
	 */
	getConfigFields(){
		return this.configFields;
	}

	/**
	 * (re)Set the config fields
	 *
	 * @param {Object} configFields
	 */
	setConfigFields(configFields){
		this.configFields = configFields;
	}

	/**
	 * Get the config values
	 *
	 * @returns {Object}
	 */
	getConfigValues(){
		return this.configValues;
	}

	/**
	 * (re)Set the config values
	 *
	 * @param {Object} configValues
	 */
	setConfigValues(configValues){
		this.configValues = configValues;
	}

	/**
	 * Checks if default config fields are available for class instance.
	 *
	 * Specifically checks if a static property defaultConfigFields of class instance default is an object.
	 * @returns {boolean}
	 */
	hasDefaultConfigFields(){
		return 'object' === typeof this.constructor.defaultConfigFields;
	}
}