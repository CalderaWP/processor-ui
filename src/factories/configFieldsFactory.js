import {configFieldFactory} from './configFieldFactory';
import {validation} from '@caldera-labs/components';

/**
 * Prepare a collection of config fields
 *
 * Optionally merges with default config fields
 * Optionally sets values from a Map of config values
 *
 * @param {Object} configFields Config fields for processor. Indexed by config ID.
 * @param {Object} configFieldsDefaults Optional. Config field defaults for processor. Indexed by config ID.
 * @returns {Object}
 */
export const configFieldsFactory = (configFields, configFieldsDefaults = {}  ) => {
	Object.keys( configFields ).forEach( configFieldId => {
		if( configFieldsDefaults.hasOwnProperty(configFieldId)){
			configFields[configFieldId] = configFieldFactory(
				configFields[configFieldId],
				configFieldsDefaults[configFieldId]
			);

		}
	});

	Object.keys( configFieldsDefaults ).forEach( configDefaultId => {
		if( ! configFields.hasOwnProperty(configDefaultId)){
			configFields[configDefaultId] = configFieldsDefaults[configDefaultId];
		}
	});

	Object.keys( configFields ).forEach(configFieldId  => {
		configFields[configFieldId].id = configFieldId;
		configFields[configFieldId].ID = configFieldId;
		configFields[configFieldId] = validation.addAutomaticValidators(configFields[configFieldId]);
	});

	return configFields;
};
