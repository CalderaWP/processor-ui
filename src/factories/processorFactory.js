import {emailDefaultConfigFields} from '../processors/emailDefaultConfigFields';
import {Processor} from '../processors/Processor';
import {configFieldsFactory} from './configFieldsFactory';

export const processorTypes = new Map()
	.set('email', {
		defaultConfigFields: emailDefaultConfigFields
	} );

processorTypes.set( 'redirect', {
	url: {
		'label': 'Redirect Url',
		'desc': 'URL to redirect to',
		'type': 'text',
	},
});

export const processorFactory = (type,configFields,configValues ) =>  {
	let processor = new Processor();
	if( processorTypes.has( type ) ){
		const configFieldDefaults = processorTypes.get( type ).hasOwnProperty( 'defaultConfigFields' )
			? processorTypes.get( type ).defaultConfigFields
			: {};
		processor.setConfigFields( configFieldsFactory(configFields,configFieldDefaults));

	}else{
		processor.setConfigFields( configFields );
	}


	processor.setConfigValues( configValues);
	return processor;

};