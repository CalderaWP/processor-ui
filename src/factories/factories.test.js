import {configFieldFactory} from './configFieldFactory';
import {configFieldsFactory} from './configFieldsFactory';
import {processorFactory} from './processorFactory';
import {getConfigFieldDefaultValue, getConfigFieldValueOrDefault} from './util';
import {Processor} from '../processors/Processor';

describe('Processor factories', () => {
	//A mock library is needed!
	const textFieldConfig = {
		'label': 'Tags',
		'description': 'Comma separated list of tags.',
		'type': 'text',
		'default': false
	};

	const defaultSequenceId = 42;
	const hiddenFieldConfig = {
		'type': 'hidden',
		'label': 'Sequence ID',
		'default': defaultSequenceId
	};

	const configFields = {
		tags: textFieldConfig,
		sequenceId: hiddenFieldConfig
	};

	describe('configFieldFactory', () => {
		it('Parses defaults', () => {
			expect(configFieldFactory({
				description: 'Hi Roy'
			}, hiddenFieldConfig).description).toEqual('Hi Roy');
		});

		it('Sets value from default', () => {
			expect(configFieldFactory({
				description: 'Hi Roy'
			}, hiddenFieldConfig).value).toEqual(defaultSequenceId);
		});

		it('Sets configField.default from configFieldDefaults.default when needed', () => {
			expect(configFieldFactory(
				{
					description: 'Hi Roy'
				},
				{
					default: 'zooms'
				}
			).default).toEqual('zooms');
		});


		it('Sets missing value from configFieldDefaults.default when needed', () => {
			expect(configFieldFactory(
				{
					description: 'Hi Roy'
				},
				{
					default: 'zooms'
				}
			).value).toEqual('zooms');
		});
	});

	describe('configFieldsFactory', () => {
		it('Adds the fields ', () => {
			const processorConfigFields = configFieldsFactory(configFields);
			expect(processorConfigFields.has('tags')).toEqual(true);
			expect(processorConfigFields.has('sequenceId')).toEqual(true);
		});

		it('Adds IDs to the fields ', () => {
			const processorConfigFields = configFieldsFactory(configFields);
			expect(processorConfigFields.get('tags').ID).toEqual('tags');
			expect(processorConfigFields.get('sequenceId').ID).toEqual('sequenceId');
		});

		describe('Parses defaults properly', () => {
			it('Adds missing configFields', () => {
				const processorConfigFields = configFieldsFactory({
					tags: textFieldConfig,
				}, configFields);
				expect(processorConfigFields.has('sequenceId')).toEqual(true);
				expect(processorConfigFields.get('sequenceId').label).toEqual(hiddenFieldConfig.label);
			});

			it('Overrides defaults', () => {
				const newTagsField = Object.assign({}, textFieldConfig);
				newTagsField.type = 'select';
				const processorConfigFields = configFieldsFactory({
					tags: newTagsField,
				}, configFields);
				expect(processorConfigFields.get('tags').type).toEqual('select');

			});

			it('Adds missing config field defaults', () => {
				const newTagsField = {
					type: 'select'
				};
				const processorConfigFields = configFieldsFactory({
					tags: newTagsField,
				}, configFields);
				expect(processorConfigFields.has('tags')).toEqual(true);
				expect(processorConfigFields.get('tags').label).toEqual('Tags');
				expect(processorConfigFields.get('tags').default).toEqual(false);

			});


		});

		it('Sets fieldConfig.value to default when no field value is present ', () => {
			const processorConfigFields = configFieldsFactory(configFields, {}, new Map().set('tags', 'roy'));
			expect(processorConfigFields.get('sequenceId').value).toEqual(42);

		});

		it('Sets fieldConfig.value if configValues has a value to set', () => {
			const processorConfigFields = configFieldsFactory({
				tags: textFieldConfig,
			}, configFields, new Map().set('tags', 'roy'));
			expect(processorConfigFields.get('tags').value).toEqual('roy');

		});

		it('Adds in keys from default not in config', () => {
			const thisTestTextFieldConfig = {
				'label': 'r3s',

			};
			const thisTestConfigFields = {
				tags: thisTestTextFieldConfig,
				sequenceId: hiddenFieldConfig
			};

			const processorConfigFields = configFieldsFactory(thisTestConfigFields, configFields);
			expect(processorConfigFields.get('tags').label).toEqual(thisTestTextFieldConfig.label);
			expect(processorConfigFields.get('tags').type).toEqual('text');

		});

		describe('Priority of setting value', () => {
			const defaultConfigFieldValue = 'mike';
			const defaultsForThisTest = Object.assign({}, {
				tags: {
					default: 'mike',
				},
				sequenceId: hiddenFieldConfig
			});

			it('is testing with a valid mock for defaults', () => {
				expect(defaultsForThisTest.tags.default).toEqual('mike');
			});


			describe('Gives top priority to fieldValues[configFieldId]', () => {
				const processorConfigFields = configFieldsFactory(
					{
						tags: textFieldConfig,
						sequenceId: hiddenFieldConfig
					},
					Object.create(defaultsForThisTest),
					new Map().set('tags', 'roy')
				);
				expect(processorConfigFields.get('tags').value).not.toEqual(defaultConfigFieldValue);
			});

			describe('Uses config.Value as 2nd priority for setting value', () => {
				const processorConfigFields = configFieldsFactory(
					{
						tags: {
							default: 't1',
							label: 'Tags2'
						},
						sequenceId: hiddenFieldConfig
					},
					Object.create(defaultsForThisTest),
					new Map() //no values
				);

				it('Sets the value on fieldConfig from configFields', () => {
					expect(processorConfigFields.get('tags').value).toEqual('t1');
				});
			});

			describe('Uses configField.default as 3rd priority for setting value', () => {
				const processorConfigFields = configFieldsFactory(
					{
						tags: {
							label: 'Tags3',
							default: 'torpedo'
						},
						sequenceId: hiddenFieldConfig
					},
					Object.create(defaultsForThisTest),
					new Map() //no values
				);

				it('Sets the default on fieldConfig from defaultConfigFieldValue', () => {
					expect(processorConfigFields.get('tags').value).toEqual('torpedo');
				});
			});

			describe('Uses defaultConfigField.default as 4th priority for setting value', () => {

				it('Has the right default in mock', () => {
					expect(defaultsForThisTest.tags.default).toEqual('mike');
				});
				it('Uses getConfigFieldValueOrDefault as intended', () => {
					expect(getConfigFieldDefaultValue({default: 'mike'})).toEqual('mike');
					expect(getConfigFieldValueOrDefault({default: 'mike'})).toEqual('mike');
				});

				it('Sets the default on fieldConfig from defaultConfigFieldValue', () => {
					const processorConfigFields = configFieldsFactory(
						{
							tags: {
								label: 'Tags4',
							},
							sequenceId: hiddenFieldConfig
						},
						{
							tags: {
								default: 'mike',
							}
						},
						new Map() //no values
					);
					expect(processorConfigFields.get('tags').value).toEqual(defaultsForThisTest.tags.default);
				});
			});
		});
	});

	describe('Main processor factory', () => {
		describe('Matches type', () => {
			it('Returns correct class instance', () => {
				expect(processorFactory('robots', {})).toBeInstanceOf(Processor);
			});

			it('Returns correct class instance for email processor', () => {
				expect(processorFactory('email', {})).toBeInstanceOf(Processor);
			});

			it('Parse defaults', () => {
				const configFields = {
					fromEmail: {
						label: 'Butter Emails'
					}
				};
				let processor = processorFactory('email', configFields);

				expect(processor.getConfigFields().get( 'fromEmail' ).hasOwnProperty('label')).toEqual(true);
			});
		});
	});
});