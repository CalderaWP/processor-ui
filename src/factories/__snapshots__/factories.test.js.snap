// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Processor factories configFieldsFactory sets the defaults 1`] =
`"{\\"fromName\\":{\\"label\\":\\"From Name\\",\\"type\\":\\"email\\",\\"value\\":\\"Roy\\",\\"id\\":\\"fromName\\",\\"ID\\":\\"fromName\\",\\"validators\\":[]}}"`;
